Change instructions on how to track deprecated items for removal and add a few more links

aescolar opened the pull request:
Following the discussion in discord: https://discord.com/channels/720317445772017664/733037890514321419/1272496160308723754

Let's change the docs, so we tell developers to add to one GitHub issue tracking all deprecated APIs meant to be removed, so the process is simpler and hopefully developers do it.

Let's also add a few references from the options they use so we remind developers to do so.

aescolar committed:
docs: api_lifecycle: Change where to list deprecated items

When deprecating APIs,
let's tell users to add to one GitHub issue tracking all
deprecated APIs meant to be removed,
so the process is simpler and hopefully developers do it.

Signed-off-by: Alberto Escolar Piedras <alberto.escolar.piedras@nordicsemi.no>

aescolar committed:
Kconfig: Link to deprecation instructions

Let's link from the deprecated instructions.
To increase visibility and hopefully avoid developers forgetting to
do so.

Signed-off-by: Alberto Escolar Piedras <alberto.escolar.piedras@nordicsemi.no>

aescolar committed:
cmake: deprecated.cmake: Link to deprecation instructions

Let's link to the deprecated instructions.
To increase visibility and hopefully avoid developers forgetting to
do so.

Signed-off-by: Alberto Escolar Piedras <alberto.escolar.piedras@nordicsemi.no>

tomi-font commented on doc/develop/api/api_lifecycle.rst:
```diff
@@ -236,9 +236,8 @@ The following are the requirements for deprecating an existing API:
-  - Create a GitHub issue to track the removal of the deprecated API, and
-    add it to the roadmap targeting the appropriate release
-    (in the example above, 1.16).
+  - Add an entry in the `GitHub issue <https://github.com/zephyrproject-rtos/zephyr/issues/76943>`_
+    tracking removal of deprecated APIs.
```
You mean that people edit the main post? If so, will everyone have the rights to do that?

henrikbrixandersen commented on doc/develop/api/api_lifecycle.rst:
```diff
@@ -236,9 +236,8 @@ The following are the requirements for deprecating an existing API:
-  - Create a GitHub issue to track the removal of the deprecated API, and
-    add it to the roadmap targeting the appropriate release
-    (in the example above, 1.16).
+  - Add an entry in the `GitHub issue <https://github.com/zephyrproject-rtos/zephyr/issues/76943>`_
+    tracking removal of deprecated APIs.
```
I think we should have one GitHub issue per release, documenting which deprecated functionality is to be removed prior to that release instead of having a singleton issue, which needs editing all the time.

The benefit of having one GitHub issue per release for tracking this, is that each issue can be added to the given milestone and easy tracked during release meetings, release checklists and so forth.

aescolar replied:
Having an issue per release we cannot link at it from the docs or code like this PR is doing.
Also somebody needs to create that issue for each release, and that person is very likely not going to find a milestone ready.
So I'd expect what will happen if we do that is that the first few items which will be deprecated in each release will not be noted anywhere because there were too many steps.
Just like now nobody is creating an issue for each deprecation and setting it in the release plan to the right release.

aescolar replied:
If you wanted, we could have developers add to this issue, and at the start of each release cut out the section for that release into a separate issue.
(Developers should normally only add 2 releases ahead)

dkalowsk replied:
Correct me if I'm wrong here, I believe the proposed solution was to use the single issue to contain a list of the next few releases and the expected deprecation items for them.  At release time, one of the checklist items for the release managers is to copy the issue and increment the release version it tracks (i.e. drop 4.0 and add 4.3), and then bump this file version forward on the issue to track.  Does that sound correct?

aescolar replied:
The PR has been changed to what was discussed and agreed in yesterday's release meeting.
@henrikbrixandersen @dkalowsk please re-review

dkalowsk commented on Kconfig.zephyr:
```diff
@@ -949,6 +949,8 @@ config DEPRECATED
 	  Symbol that must be selected by a feature or module if it is
 	  considered to be deprecated.
+	  When adding this to an option, remember to add a mention in the GitHub issue tracking the
+	  removal of deprecated APIs/options: https://github.com/zephyrproject-rtos/zephyr/issues/76943
```
This creates now 4 places to go update at release time.  It's feeling more like we should just have a "deprecation_process.rst" that all of these can link to.

aescolar committed:
docs: api_lifecycle: Use a better release as example for deprecation

1.16 was not meant to be a release, and 1.14 was long ago.
Let's use as example something much more recent.

Signed-off-by: Alberto Escolar Piedras <alberto.escolar.piedras@nordicsemi.no>

aescolar committed:
toolchain/gcc.h: Link to instructions on deprecation

Let's link from the deprecated instructions,
To increase visibility and hopefully avoid developers forgetting to
do so.

Signed-off-by: Alberto Escolar Piedras <alberto.escolar.piedras@nordicsemi.no>

jukkar commented on Kconfig.zephyr:
```diff
@@ -949,6 +949,8 @@ config DEPRECATED
 	help
 	  Symbol that must be selected by a feature or module if it is
 	  considered to be deprecated.
+	  When adding this to an option, remember to follow the instructions in
+	  https://docs.zephyrproject.org/latest/develop/api/api_lifecycle.html#deprecated
```
Could this information printed during compilation too so that developer remembers to add the information to the issue?

henrikbrixandersen replied:
The compilation warnings are not targeting the developer doing the deprecation, but rather the users affected by the deprecation. I don't think this message should be included in the compilation warning.
